### PR TITLE
[4.1] Field group site display enhancement

### PR DESCRIPTION
**Pull request by ceford**
Pull Request for Issue #31385 and related to #31396  .

### Summary of Changes

Code to customise site display of Field Groups

### Testing Instructions

Create some article fields that are not in a field group.
Edit an article and populate those fields.

Test the default behaviour.

Create one or more field groups, the example I use is Ingredients and Procedure.
The PR has added some extra fields to the Options tab of the Field Group data entry form:
	Select Show Title = Yes
	Select a render tag.
Add some fields to those groups.
Edit an article and populate the fields

### Actual result BEFORE applying this Pull Request

The default behaviour is to show all fields in a single bulleted list with no 'title'.

### Expected result AFTER applying this Pull Request

The example I use shows a Recipe with Ingredients as a bulleted list and the Procedure as a numbered list, with titles.

image

### Documentation Changes Required

Help screen will need revision

### CODE REVIEW

File: administrator/components/com_fields/forms/group.xml [hunk 1]
```diff
@@ -152,6 +152,59 @@
...
+				<option value="1">JYES</option>
+			</field>
+
+			<field
+				name="title_tag"
+				type="list"
+				label="COM_FIELDS_GROUP_TITLE_TAG_LABEL"
+				default="ul"
+				>
+				<option value="p">P</option>
```
Comment (brianteeman) [inline administrator/components/com_fields/forms/group.xml]: lowercase please

File: administrator/components/com_fields/forms/group.xml [hunk 2]
```diff
@@ -152,6 +152,59 @@
...
+				default="mb-0"
+			/>
+
+			<field
+				name="render_tag"
+				type="list"
+				label="COM_FIELDS_GROUP_RENDER_TAG_LABEL"
+				default="ul"
+				>
+				<option value="ul">UL</option>
```
Comment (brianteeman) [inline administrator/components/com_fields/forms/group.xml]: lowercase please

**Comment by brianteeman** [comment]
Warning: Use of undefined constant none - assumed 'none' (this will throw an Error in a future version of PHP) in C:\htdocs\joomla-cms\components\com_fields\layouts\fields\render.php on line 91

**Comment by brianteeman** [comment]
If you dont enter a class then it should not output an empty class. Currently you get for example `<ul class="">`

**Comment by richard67** [comment]
@ceford Please use tabs and not spaces to indent code, see PHP CS errors reported by drone: [https://ci.joomla.org/joomla/joomla-cms/37579/1/6](https://ci.joomla.org/joomla/joomla-cms/37579/1/6)

**Comment by brianteeman** [comment]
Depending on the options you select they are either ignored or they generate invalid mark

You cannot have the following markup
```
<p>title</p>
<li>field1</li>
<li>field2</li>
```

But it gets much worse with other combinations

**Comment by brianteeman** [comment]
Not sure what the show/hide title is supposed to do but I couldnt see it doing anything

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 29a934a69b55e503458064c334324c4c3559b1cb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

**Comment by brianteeman** [comment]
Thanks for the fixes - getting better

image

This generates the following

image

I get why you made it do that - p and dl would be invalid but I you should not be able to select an invalid option or combination in the first place. Otherwise users will think it is broken

It's the same when you set title to hide on a dl. The title is not hidden because it would be invalid

File: administrator/components/com_fields/forms/group.xml [hunk 3]
```diff
@@ -152,6 +152,59 @@
...
+				>
+				<option value="0">JNO</option>
+				<option value="1">JYES</option>
+			</field>
+
+			<field
+				name="title_tag"
+				type="list"
+				label="COM_FIELDS_GROUP_TITLE_TAG_LABEL"
+				default="ul"
```
Comment (brianteeman) [inline administrator/components/com_fields/forms/group.xml]: the default value of `ul` is not present in the list of options?

**Comment by ceford** [comment]
Thanks to @brianteeman for obliging me to think about the form behaviour. After wondering how to compose some javascript and poking around I discovered that all I had to do was add a field attribute to the form xml: showon="show_title:1[AND]render_tag:ul,ol,p,div" which I can only describe as absolutely marvelous. My complements to the javascript developers. I wonder if this sort of stuff is documented somewhere?

**Comment by brianteeman** [comment]
> I wonder if this sort of stuff is documented somewhere?

https://docs.joomla.org/Form_field#Showon

**Comment by ashvini77** [comment]
I have tested this item :red_circle: unsuccessfully on 092e829e4dc5051af06c6c71b06ae2cb1e8f5f5e<br /><br />https://tinyurl.com/y5r5ys4z<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

**Comment by ghost** [comment]
@ashvini77 you can add a screenshot here so user don't have to open a link.

**Comment by punambaravkar** [comment]
I have tested this item :white_check_mark: successfully on 092e829e4dc5051af06c6c71b06ae2cb1e8f5f5e<br /><br />Yes, After adding the patch, it works correctly <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

**Comment by ashvini77** [comment]
screen shot 2020-12-05 at 07 35 22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

**Comment by vaibhavsTekdi** [comment]
I have tested this item :white_check_mark: successfully on 092e829e4dc5051af06c6c71b06ae2cb1e8f5f5e<br /><br />It works only if when I save that field group after applying the patch again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

**Comment by ceford** [comment]
The bug occurs when you set up the test groups before applying the patch. To fix it, apply the patch and then open each Field Group in turn and save. That updates the parameters. I have change the code for backward compatibility purposes. You still need to save the Field Groups with the new features to store them. Please test again.

File: components/com_fields/layouts/fields/render.php [hunk 4]
```diff
@@ -69,14 +73,68 @@
 		continue;
 	}
 
-	$output[] = '<li class="field-entry ' . $class . '">' . $content . '</li>';
+	// check gparams for backward compatibility
+	$gparams = json_decode($field->group_params);
+	
+	if (isset($field->group_title) && isset($gparams->render_tag))
```
Comment (Quy) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
	$gparams = json_decode($field->group_params);

	if (isset($field->group_title) && isset($gparams->render_tag))
```

Comment (ceford) [reply components/com_fields/layouts/fields/render.php]: Fixed - removed trailing white space. I recently installed a new version of Eclipse PDT and did not get the settings right!

**Comment by ceford** [comment]
Pity this is stuck in limbo. All tests have passed and the adverse comments resulted from testing procedure.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31408">issues.joomla.org/tracker/joomla-cms/31408</a>.</sub>

**Comment by bembelimen** [comment]
Nice feature, for sure something for 4.1

**Comment by ceford** [comment]
Why put this off for 4.1? It was always intended for 4.0.

File: components/com_fields/layouts/fields/render.php [hunk 5]
```diff
@@ -69,14 +73,68 @@
...
+			default:
+				$tag = 'li';
+		}
+		$list_type[$field->group_title] = $gparams->render_tag;
+		$list_type_class[$field->group_title] = $gparams->render_class;
+		$show_title[$field->group_title] = $gparams->show_title;
+		$title_tag[$field->group_title] = $gparams->title_tag;
+		$title_class[$field->group_title] = $gparams->title_class;
+		$output[$field->group_title][] = '<' . $tag  . ' class="field-entry ' . $class . '">' . $content . '</' . $tag  . '>';
+	} else {
```
Comment (infograf768) [inline components/com_fields/layouts/fields/render.php]: }
else
{

File: components/com_fields/layouts/fields/render.php [hunk 6]
```diff
@@ -69,14 +73,68 @@
...
-</ul>
+
+foreach ($list_type as $title => $list_type)
+{
+	if ($show_title[$title] && $list_type != 'dl')
+	{
+		$class = empty($title_class[$title]) ? '' : ' class="' . $title_class[$title] . '"';
+		echo '<' . $title_tag[$title] . $class . '>' . $title . '</' . $title_tag[$title] . '>' . "\n";
+	}
+	if (!($list_type == 'p' || $list_type == 'div'))
```
Comment (infograf768) [inline components/com_fields/layouts/fields/render.php]: add blank line before if

File: components/com_fields/layouts/fields/render.php [hunk 7]
```diff
@@ -69,14 +73,68 @@
...
+	{
+		$class = empty($list_type_class[$title]) ? '' : ' class="' . $list_type_class[$title]  . '"';
+		echo '<' . $list_type . $class . '>';
+	}
+	if ($list_type == 'dl')
```
Comment (infograf768) [inline components/com_fields/layouts/fields/render.php]: add blank line before if

File: components/com_fields/layouts/fields/render.php [hunk 8]
```diff
@@ -69,14 +73,68 @@
...
+	{
+		$class = empty($list_type_class[$title]) ? '' : ' class="' . $list_type_class[$title] . '"';
+		echo '<dt'. $class . '>' . $title . '</dt>' . "\n";
+	}
+	echo implode("\n", $output[$title]);
+	if (!($list_type == 'p' || $list_type == 'div'))
```
Comment (infograf768) [inline components/com_fields/layouts/fields/render.php]: add blank line before if

**Comment by richard67** [comment]
@wilsonge Your opinion on this? 4.0 or 4.1?

**Comment by ceford** [comment]
This came up as a gripe in the forum because switching custom fields from dl to ul in J4 ruined perfectly good article appearance. Put it off to 4.1 and expect a lot more gripes!

**Comment by brianteeman** [comment]
That gripe was already resolved. This adds new functionality

**Comment by bembelimen** [comment]
> 
> 
> That gripe was already resolved. This adds new functionality

Exactly. It's a really good enhancement but we should stop adding new stuff into 4.0 to get it released. But with a bit of clean up, we'll have it in 4.1.

File: components/com_fields/layouts/fields/render.php [hunk 9]
```diff
@@ -69,14 +73,73 @@
...
+			}
+			case 'p':
+			case 'div':
+			{
+				$tag = $gparams->render_tag;
+				break;
+			}
+			default:
+				$tag = 'li';
+		}
```
Comment (PhilETaylor) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
		switch ($gparams->render_tag)
		{
			case 'dl':
				$tag = 'dd';
				break;
			case 'p':
			case 'div':
				$tag = $gparams->render_tag;
				break;
			default:
				$tag = 'li';
		}
```

File: components/com_fields/layouts/fields/render.php [hunk 10]
```diff
@@ -50,6 +50,10 @@
 }
 
 $output = array();
+$list_type = array();
+$show_title = array();
+$title_tag = array();
+$title_class = array();
```
Comment (PhilETaylor) [inline components/com_fields/layouts/fields/render.php]: Its 2021.

```suggestion
$output = [];
$list_type = [];
$show_title = [];
$title_tag = [];
$title_class = [];
```

File: components/com_fields/layouts/fields/render.php [hunk 11]
```diff
@@ -69,14 +73,73 @@
...
+
+foreach ($list_type as $title => $list_type)
+{
+	if ($show_title[$title] && $list_type != 'dl')
+	{
+		$class = empty($title_class[$title]) ? '' : ' class="' . $title_class[$title] . '"';
+		echo '<' . $title_tag[$title] . $class . '>' . $title . '</' . $title_tag[$title] . '>' . "\n";
+	}
+
+	if (!($list_type == 'p' || $list_type == 'div'))
```
Comment (PhilETaylor) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
	if (!($list_type === 'p' || $list_type === 'div'))
```

File: components/com_fields/layouts/fields/render.php [hunk 12]
```diff
@@ -69,14 +73,73 @@
...
+	{
+		$class = empty($list_type_class[$title]) ? '' : ' class="' . $list_type_class[$title]  . '"';
+		echo '<' . $list_type . $class . '>';
+	}
+
+	if ($list_type == 'dl')
```
Comment (PhilETaylor) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
	if ($list_type === 'dl')
```

File: components/com_fields/layouts/fields/render.php [hunk 13]
```diff
@@ -69,14 +73,73 @@
...
+	{
+		$class = empty($list_type_class[$title]) ? '' : ' class="' . $list_type_class[$title] . '"';
+		echo '<dt'. $class . '>' . $title . '</dt>' . "\n";
+	}
+	echo implode("\n", $output[$title]);
+
+	if (!($list_type == 'p' || $list_type == 'div'))
```
Comment (PhilETaylor) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
	if (!($list_type === 'p' || $list_type === 'div'))
```

**Comment by Krshivam25** [comment]
@ceford Can you please commit this changes so we can test this again.

**Comment by ceford** [comment]
Any chance someone else can pick this up? Right now I don't know what to do - too long ago for may failing memory.

**Comment by ceford** [comment]
I have made a complete pig's ear of the changes requested by Phil. My local repo is in a mess and I don't know how to fix it or revoke my latest commit. Annoyed with myself and ...

**Comment by ceford** [comment]
I am going to have to redo this from scratch starting from a new repo clone.

**Pull request closed without merging**